### PR TITLE
Chapter 12 - Password reset

### PR DESCRIPTION
**Pull request by EvonneNgXinYi**
Reset the password when user forgot it

### CODE REVIEW

**Comment by EvonneNgXinYi** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?
+      @user.errors.add :password, "can't be empty"
+      render :edit
+    elsif @user.update(user_params)
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    elsif @user.update user_params
```

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      log_in @user
+      @user.update_attribute :reset_digest, nil
+      flash[:success] = t ".success_psword_update"
+      redirect_to @user
+    else
+      flash[:danger] = t ".fail_ps_reset_mess"
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
      flash.now[:danger] = t ".fail_ps_reset_mess"
```

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      render :edit
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:password, :password_confirmation)
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    params.require(:user).permit :password, :password_confirmation
```

File: app/views/user_mailer/password_reset.text.erb [hunk 4]
```diff
@@ -1,3 +1,5 @@
 <h1><%= t ".heading" %></h1>
-
-<%= @greeting %> <%= t ".desc_b" %>
+<%= t ".reset_title" %>
+<%= edit_password_reset_url(id: @user.reset_token, email: @user.email) %>
```
Comment (chungpv-1008) [inline app/views/user_mailer/password_reset.text.erb]:
```suggestion
<%= edit_password_reset_url id: @user.reset_token, email: @user.email %>
```

Comment (EvonneNgXinYi) [reply app/views/user_mailer/password_reset.text.erb]: done

File: app/controllers/password_resets_controller.rb (hunk 3 above)
Comment (EvonneNgXinYi) [reply app/controllers/password_resets_controller.rb]: done

**Comment by EvonneNgXinYi** [comment]
ready

**Comment by EvonneNgXinYi** [comment]
ready

**Comment by EvonneNgXinYi** [comment]
ready

**Comment by EvonneNgXinYi** [comment]
ready

File: app/helpers/password_resets_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
```
Comment (ngocvt-0484) [inline app/helpers/password_resets_helper.rb]: delete this file because you not use

Comment (EvonneNgXinYi) [reply app/helpers/password_resets_helper.rb]: done

**Comment by EvonneNgXinYi** [comment]
**ready**

**Comment by vanvtt-0952** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    else
+      flash.now[:danger] = t ".fail_acct_sent"
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:23]:
```suggestion
    if params[:user][:password].blank?
```

**Review by saobv-2223 — COMMENTED** [review]
Small bug, fix it and merge

**Pull request merged**